ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixClusterManager.java

```diff
@@ -837,5 +830,461 @@ long getAllocatedRawCapacity() {
     long getAllocatedUsableCapacity() {
       return clusterWideAllocatedUsableCapacityBytes.get();
     }
+
+    /**
+     * Gets the {@link RoutingTableSnapshot} for a given data center or entire cluster.
+     * @param dcName data center name.
+     * @return {@link RoutingTableSnapshot} for a given data center. If the input data center name is null or if using
+     * aggregated view, returns the routing table snapshot of entire cluster.
+     */
+    public RoutingTableSnapshot getRoutingTableSnapshot(String dcName) {
+      return dcToRoutingTableSnapshotRef.get(dcName).get();
+    }
+  }
+
+  /**
+   * A more dynamic implementation of {@link HelixClusterChangeHandler} which supports adding new nodes/partitions at
+   * runtime. It is also able to absorb replica location changes in cluster.
+   */
+  class HelixClusterChangeHandler
+      implements DataNodeConfigChangeListener, LiveInstanceChangeListener, IdealStateChangeListener,
+                 RoutingTableChangeListener {
+    private final String dcName;
+    private final Object notificationLock = new Object();
+    private final Consumer<Exception> onInitializationFailure;
+    private final HelixClusterManager.HelixClusterManagerCallback helixClusterManagerCallback;
+    private final CountDownLatch routingTableInitLatch = new CountDownLatch(1);
+    private final List<ClusterMapChangeListener> clusterMapChangeListeners = new ArrayList<>();
+    private volatile boolean instanceConfigInitialized = false;
+    private volatile boolean liveStateInitialized = false;
+    private volatile boolean idealStateInitialized = false;
```

Comment:
`HelixClusterChangeHandler` is same as `DynamicClusterChangeHandler`. The only difference is that data structures holding cluster information have been moved to `HelixClusterManager`.